rclcpp/src/rclcpp/signal_handler.cpp

```diff
@@ -161,82 +205,41 @@ SignalHandler::~SignalHandler()
 }
 
 SignalHandler::signal_handler_type
-SignalHandler::set_signal_handler(
-  int signal_value,
-  const SignalHandler::signal_handler_type & signal_handler)
+SignalHandler::get_old_signal_handler(int signum)
 {
-  bool signal_handler_install_failed;
-  SignalHandler::signal_handler_type old_signal_handler;
+  if (SIGINT == signum) {
+    return old_sigint_handler_;
+  } else if (SIGTERM == signum) {
+    return old_sigterm_handler_;
+  }
 #if defined(RCLCPP_HAS_SIGACTION)
-  ssize_t ret = sigaction(signal_value, &signal_handler, &old_signal_handler);
-  signal_handler_install_failed = (ret == -1);
+  SignalHandler::signal_handler_type ret;
+  memset(&ret, 0, sizeof(ret));
+  sigemptyset(&ret.sa_mask);
+  ret.sa_handler = SIG_DFL;
+  return ret;
 #else
-  old_signal_handler = std::signal(signal_value, signal_handler);
-  signal_handler_install_failed = (old_signal_handler == SIG_ERR);
+  return SIG_DFL;
 #endif
-  if (signal_handler_install_failed) {
-    char error_string[1024];
-    rcutils_strerror(error_string, sizeof(error_string));
-    auto msg =
-      "Failed to set SIGINT signal handler (" + std::to_string(errno) + "): " + error_string;
-    throw std::runtime_error(msg);
-  }
-
-  return old_signal_handler;
 }
 
 void
 SignalHandler::signal_handler_common()
 {
-  signal_received_.store(true);
+  auto & instance = SignalHandler::get_global_signal_handler();
+  instance.signal_received_.store(true);
   RCLCPP_DEBUG(
     get_logger(),
-    "signal_handler(): SIGINT received, notifying deferred signal handler");
-  notify_signal_handler();
+    "signal_handler(): notifying deferred signal handler");
+  instance.notify_signal_handler();
 }
 
-#if defined(RCLCPP_HAS_SIGACTION)
-void
-SignalHandler::signal_handler(int signal_value, siginfo_t * siginfo, void * context)
-{
-  RCLCPP_INFO(get_logger(), "signal_handler(signal_value=%d)", signal_value);
-
-  if (old_signal_handler_.sa_flags & SA_SIGINFO) {
-    if (old_signal_handler_.sa_sigaction != NULL) {
-      old_signal_handler_.sa_sigaction(signal_value, siginfo, context);
-    }
-  } else {
-    if (
-      old_signal_handler_.sa_handler != NULL &&  // Is set
-      old_signal_handler_.sa_handler != SIG_DFL &&  // Is not default
-      old_signal_handler_.sa_handler != SIG_IGN)  // Is not ignored
-    {
-      old_signal_handler_.sa_handler(signal_value);
-    }
-  }
-
-  signal_handler_common();
-}
-#else
-void
-SignalHandler::signal_handler(int signal_value)
-{
-  RCLCPP_INFO(get_logger(), "signal_handler(signal_value=%d)", signal_value);
-
-  if (old_signal_handler_) {
-    old_signal_handler_(signal_value);
-  }
-
-  signal_handler_common();
-}
-#endif
-
 void
 SignalHandler::deferred_signal_handler()
 {
   while (true) {
     if (signal_received_.exchange(false)) {
```

Comment:
@ivanpauno meta: we could probably forward the signal number through an atomic integer instead.

Reply:
you could get two signals before the deferred signal handler starts running, so it really doesn't help much.

Reply:
Fair point. In which case differentiated handling is not possible. Then, `shutdown_on_sigint` --> `shutdown_on_signal`.

Reply:
Renaming sounds fine to me, though it's an unnecessary breaking change.

Reply:
>  though it's an unnecessary breaking change.

I disagree, but perhaps others can chime in. @wjwwood ?

Reply:
Hmm, I don't mind, we could also alias them and deprecate the old one through a tick-tock cycle. I agree that "just changing it" it probably unnecessarily breaking.

Reply:
> you could get two signals before the deferred signal handler starts running, so it really doesn't help much.

The only important thing is that if you get sigterm and sigint you notice both (regardless of how many times they occur), so you could just have two atomic bools to check here, one for sigint and one for sigterm.

Reply:
> The only important thing is that if you get sigterm and sigint you notice both (regardless of how many times they occur), so you could just have two atomic bools to check here, one for sigint and one for sigterm.

yes, I think that's possible.
It's a bit tricky though